op-node/sources/l1_client.go

```diff
@@ -115,3 +119,27 @@ func (s *L1Client) L1BlockRefByHash(ctx context.Context, hash common.Hash) (eth.
 	s.l1BlockRefsCache.Add(ref.Hash, ref)
 	return ref, nil
 }
+
+// L1HotShotCommitmentsFromHeight returns an array of HotShot commitments to sequencer blocks
+// This is used in the derivation pipeline to validate sequencer batches in Espresso mode
+func (s *L1Client) L1HotShotCommitmentsFromHeight(firstBlockHeight uint64, numHeaders uint64, hotshotAddr common.Address) ([]espresso.NmtRoot, error) {
+	var comms []espresso.NmtRoot
+	client := ethclient.NewClient(s.client.RawClient())
+	hotshot, err := hotshot.NewHotshot(hotshotAddr, client)
+	if err != nil {
+		return nil, err
+	}
+	for i := 0; i < int(numHeaders); i++ {
+		var height big.Int
+		height.SetUint64(firstBlockHeight + uint64(i))
+		comm, err := hotshot.HotshotCaller.Commitments(nil, &height)
+		if err != nil {
+			return comms, err
+		}
+		root := espresso.NmtRoot{
+			Root: comm.Bytes(),
```

Comment:
We're cheating here, because the commitment bytes aren't actually an NMT root, it's the hash of the NMT root and other metadata. I think we need to either fetch the full headers from the contract calldata, or, perhaps easier, just put the full headers in the `Justification` and compute their commitment to check against these bytes. So I think this function should be changed to return the `Commitment` type.

If you want, I can do the `Justification` change as part of #53 , since I will need to refactor `Justification` for that anyways.

Reply:
Putting it in justification sounds good 👍 I can comment out the check for now

Reply:
Switched to the ```Commitment``` type and added a TODO to validate the actual hash against the full headers